evennia/objects/objects.py

```diff
@@ -612,25 +612,56 @@ def msg(self, text=None, from_obj=None, session=None, options=None, **kwargs):
                 (which can be a string or a tuple).
 
         Notes:
-            `at_msg_receive` will be called on this Object.
+            `at_msg_receive`and at_post_msg_receive will be called on this Object.
             All extra kwargs will be passed on to the protocol.
 
         """
-        # try send hooks
-        if from_obj:
-            for obj in make_iter(from_obj):
-                try:
-                    obj.at_msg_send(text=text, to_obj=self, **kwargs)
-                except Exception:
-                    logger.log_trace()
         kwargs["options"] = options
-        try:
-            if not self.at_msg_receive(text=text, from_obj=from_obj, **kwargs):
-                # if at_msg_receive returns false, we abort message to this object
-                return
-        except Exception:
-            logger.log_trace()
+        self._msg_helper_text_format(text, kwargs)
+
+        # try send hooks
+        self._msg_helper_from_obj(from_obj=from_obj, **kwargs)
+
+        if not self._msg_helper_receive(from_obj=from_obj, **kwargs):
+            return
 
+        # relay to session(s)
+        self._msg_helper_session_relay(session=session, **kwargs)
+
+        self.at_post_msg_receive(from_obj=from_obj, **kwargs)
+
+    def at_post_msg_receive(self, from_obj=None, **kwargs):
+        """
+        Overloadable hook which receives the kwargs that exist at the tail end of self.msg()'s processing.
+
+        This might be used for logging and similar purposes.
+
+        Kwargs:
+            from_obj (DefaultObject or list[DefaultObject]): The objects that sent the message.
+            **kwargs: The kwargs from the end of message, using the Evennia outputfunc format.
+        """
+        pass
```

Comment:
If this is the only method intended to be overridden, it would be more visible to move it to the end of this block, after all the private methods.